[test] packaging: add windows boxes

andyb-elastic opened the pull request:
Adds windows server 2012r2 and 2016 vagrant boxes to packaging tests.
They can only be used if IDs for their images are specified, which are
passed to gradle and then to vagrant via env variables. Adds options
to the project property `vagrant.boxes` to choose between linux and
windows boxes

Bats tests are run only on linux boxes, and portable packaging tests run
on all boxes. Platform tests are only run on linux boxes since they are
not being maintained.

For #26741

The full packaging test jobs (both periodic and pull request) should be all set up to use these boxes once this is merged.

andyb-elastic committed:
[test] packaging: add windows boxes

Adds windows server 2012r2 and 2016 vagrant boxes to packaging tests.
They can only be used if IDs for their images are specified, which are
passed to gradle and then to vagrant via env variables. Adds options
to the project property `vagrant.boxes` to choose between linux and
windows boxes

Bats tests are run only on linux boxes, and portable packaging tests run
on all boxes. Platform tests are only run on linux boxes since they are
not being maintained.

For #26741

elasticmachine commented:
Pinging @elastic/es-core-infra

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy:
```diff
@@ -59,8 +68,10 @@ class VagrantTestPlugin implements Plugin<Project> {
     @Override
     void apply(Project project) {
 
+        collectAvailableBoxes(project)
```
Since this is static maybe you should do it in a static block instead of on plugin application.

andyb-elastic replied:
Sure that makes sense. Actually general question about that (static block vs static method). I've always erred on the side of not doing stuff in static blocks when it could conceivably blow up (not sure if that applies here) because it seems more intuitive for it to blow up when it's used (in this case when the plugin is applied) than when the class is loaded. Does that make sense or is that just a weird preconception I have

nik9000 replied:
It makes some sense, yeah. static blocks are executed when the class is loaded but something like this is executed on every plugin apply. Minimally, the stack trace of the second one is going to be more useful than the first. You could certainly do the static stuff on *first* plugin apply, but that feels a little more complicated than it is worth in this scenario.

andyb-elastic replied:
That's a good point. I'll put it in a static block since it's applicable to all applications of the plugin, in contrast to some of the other configuration here which uses properties of the project

andyb-elastic replied:
This method is now non-static because of some changes I made to avoid separate plugin instances stomping on each others' configuration

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy:
```diff
@@ -287,6 +341,20 @@ class VagrantTestPlugin implements Plugin<Project> {
+                'VAGRANT_VAGRANTFILE': 'Vagrantfile',
+                'VAGRANT_PROJECT_DIR': "${project.projectDir.absolutePath}"
+        ]
+        vagrantEnvVars.putAll(VAGRANT_ENV_VARS)
```
I think it is weird to have these named the same way. Maybe look in `WINDOWS_BOXES` instead? Or name `VAGRANT_ENV_VARS` to `VAGRANT_BOX_ENV_VARS` or something.

andyb-elastic replied:
I agree, I'll make the naming more clear.

Actually originally I wrote it like

```
def vagrantEnvVars = [ ... ]
if ('windows-2016' in WINDOWS_BOXES) {
  vagrantEnvVars.put('VAGRANT_WINDOWS_2016_BOX', WINDOWS_BOXES['windows-2016'])
}
// same for windows 2012r2
return vagrantEnvVars
```

but changed it to just store the env vars directly because it seemed like an unnecessary level of indirection. What do you think?

nik9000 replied:
I'm fine with digging through the `WINDOWS_BOXES` though I might want  add something that fails if there is an unknown key in there. It'd be inconvenient to just add something to that Map and forget to update this method and have the box fail to show up mysteriously.

nik9000 reviewed (commented):
I left two questions but most of it makes sense to me.

andyb-elastic committed:
code review fixes

Get windows box IDs from environment variables in static context instead
of on plugin application

Clarify how the plugin determines the environment variables to run
vagrant with

andyb-elastic commented:
@nik9000 I pushed https://github.com/elastic/elasticsearch/pull/30402/commits/0a5b7dda068b48375c514be81a20517c7e282918

I got rid of the method that built the vagrant env vars and put that section back where it was - that method was also left over from another implementation

rjernst commented on TESTING.asciidoc:
```diff
@@ -325,26 +326,37 @@ vagrant plugin install vagrant-cachier
+* `all` - All configured boxes, i.e. all linux boxes and any configured Windows
+boxes.
+* `linux-all` - All linux boxes.
+* `windows-all` - All configured Windows boxes. If there are none configured when this
```
Thank you for avoiding leniency!

rjernst replied:
Out of curiosity, do we fail if an empty string is passed to `vagrant.boxes`? (separate issue if we don't)

andyb-elastic replied:
It does fail with an empty string, because it ends up looking for a box with the name of empty string and can't find it

rjernst commented on TESTING.asciidoc:
```diff
@@ -364,9 +376,34 @@ quality boxes available in vagrant atlas:
+These variables are required for Windows support in all gradle tasks that
+handle packaging tests. Either or both may be specified. Remember that to run tests
+on these boxes, they still need to be included in the value of `-Pvagrant.boxes`.
+When these properties are present, passing `-Pvagrant.boxes=all` will include the
```
Maybe this should fail if they are not configured? Users can run with `linux-all` to get the old behavior.

rjernst commented on TESTING.asciidoc:
```diff
@@ -364,9 +376,34 @@ quality boxes available in vagrant atlas:
+directly.
 
-* Windows Server 2012
+* `VAGRANT_WINDOWS_2012R2_BOX`
```
I think gradle properties might be better to use here. That way we could set them in `.gradle/gradle.properties`, or configure them via another gradle project within the multi project build in elasticsearch-extra. These could be `vagrant.windows-2012r2.id` and `vagrant.windows-2016.id`. They can still be passed into vagrant via env vars, but from the gradle perspective I think this would be better.

rjernst replied:
Note that to allow them to be set by a sibling project, we need to load them lazily (execution time).

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy:
```diff
@@ -386,44 +443,57 @@ class VagrantTestPlugin implements Plugin<Project> {
-                environmentVars vagrantEnvVars
-                dependsOn up, setupPackagingTest
-                finalizedBy halt
+            if (box in LINUX_BOXES) {
```
Please use a `LINUX_BOXES.contains(box)` instead of groovy's in operator, so we stay as close to java as possible (we would like to eventually switch buildSrc to all java, for performance reasons, so keeping the code as java like as possible will help with that transition). Ditto below for the other similar conditions.

rjernst reviewed (commented):
Looks great @andyb-elastic! I left a few comments.

andyb-elastic committed:
code review fixes

Set windows images ids from gradle project properties instead of
environment variables

Make vagrant test plugin configuration nonstatic so applications of the
plugin don't stomp each other

Make the plugin aware of which boxes make up the universe of
ones that we test, versus what images are actually available from the
configuration at runtime, and fail if those don't match up. Add tasks to
list all boxes and available boxes so it's clear to the user what
configuration gradle is seeing.

andyb-elastic commented:
@rjernst I pushed https://github.com/elastic/elasticsearch/pull/30402/commits/f8d3db2fd7d5aa1729e1403a5d2e4b5032235371 to address your comments

The main thing I changed was to make it keep track separately of all the boxes we test, and which ones have images available, so now it complains if any are missing. That is, it's strict in that both `vagrant.boxes=all` and `vagrant.boxes=windows-all` will fail the build if there are any windows images missing.

Note that the packaging CI job will fail for these for that reason until the workers' gradle.properties have the windows image names added, which I'll open a PR for

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy:
```diff
@@ -29,26 +31,41 @@ class VagrantTestPlugin implements Plugin<Project> {
+    ])
+
+    /** Boxes that have been supplied and are available for testing **/
+    List<String> AVAILABLE_BOXES = []
```
This and the env vars should be camelCase since they are not final.

nik9000 replied:
Can you put the non-final stuff below the final stuff? Mixing them up bothers me more than it should.

andyb-elastic replied:
Agree, makes it much more clear how it works

rjernst reviewed (approved):
LGTM, one final comment.

andyb-elastic committed:
code review fixes

Rename and relocate nonfinal, nonstatic box list members

andyb-elastic commented:
Jenkins please run gradle build tests and please run all packaging tests

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic committed:
Merge branch 'master' into feature-packaging-windows-boxes

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic commented:
Jenkins yet again run gradle build tests please

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic committed:
Merge branch 'master' into feature-packaging-windows-boxes

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic committed:
Merge branch 'master' into feature-packaging-windows-boxes
